### PR TITLE
Lookup the provided plugins on the options, throw for invalid arguments.

### LINKED ISSUE #2217: Hard to debug TypeError: arguments[i].apply is not a function
Packages version:
- webpack: 1.12.14
- tapable: 0.1.10

While editing webpack config file we met the following error:

```
<<PROJECT_PATH>>\node_modules\tapable\lib\Tapable.js:164
                arguments[i].apply(this);
                             ^

TypeError: arguments[i].apply is not a function
    at Compiler.apply (<<PROJECT_PATH>>\node_modules\tapable\lib\Tapable.js:164:16)
```

and that seemed a situation similar to issue #697 .

We were able to spot the offending change after some time, but still the message was not at all helpful. In our case, error was due to the following:

``` javascript
config.plugins = [
  // some plugins added here ...
];

// further down

config.plugins.push([
  // another plugin here, but mistakenly added as a whole array
]);
```

The fix was of course to properly add the new plugin as a single item:

``` javascript
config.plugins.push(
  // another plugin here, correctly added as a single item
);
```

Request: try to improve argument checking or anyway improve error messages.

Thanks for your time spent on this great project!

### LINKED ISSUE #6003: Hard to debug TypeError: arguments[i].apply is not a function
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

This is a means to reopening "Hard to debug TypeError: arguments[i].apply is not a function " (https://github.com/webpack/webpack/issues/2217). That issue was closed due to inactivity, but I encountered it again.

As a quick recap, I run webpack (in my case via CLI), and very soon after get:

```
PROJECT_PATH/node_modules/tapable/lib/Tapable.js:375
		arguments[i].apply(this);
               ^
TypeError: arguments[i].apply is not a function
    at Compiler.apply (PROJECT_PATH/node_modules/tapable/lib/Tapable.js:375:16)
    at webpack (PROJECT_PATH/node_modules/webpack/lib/webpack.js:33:19)
    at processOptions (PROJECT_PATH/node_modules/webpack/bin/webpack.js:335:15)
    at yargs.parse (PROJECT_PATH/node_modules/webpack/bin/webpack.js:393:2)
    at Object.Yargs.self.parse (PROJECT_PATH/node_modules/webpack/node_modules/yargs/yargs.js:533:18)
    at Object.<anonymous> (PROJECT_PATH/node_modules/webpack/bin/webpack.js:152:7)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:389:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:504:3
```

**If the current behavior is a bug, please provide the steps to reproduce.**

The issue as described in https://github.com/webpack/webpack/issues/2217 is, in my opinion, valid, as it is in fact hard to debug. I described how to reproduce it above, but additionally, I can report what was causing it—to be clear, that's what fixed it in my issue, but because the error message is ambiguous, it could also occur for other reasons.

In my case, I accidentally had pasted a non-plugin object into the `plugins` array, which I ultimately figured it out via trial and error.

**What is the expected behavior?**

A more helpful error message. Since the error happens on `arguments[i].apply(this);` perhaps reporting that it happened when parsing the `plugins` and providing the index relative the `plugins` array that's causing the error?

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

```
node 6.11.1
webpack 3.8.1
```

### PR DESCRIPTION
**Pull request by EugeneHlushko**
**What kind of change does this PR introduce?**

Schema validation for `options.plugins`.
Ensure `apply` method is available on the plugin.
Fixes #6003 
Fixes #2217

**Did you add tests for your changes?**
Yes

**If relevant, link to documentation update:**

**Summary**
"Hard to debug TypeError: arguments[i].apply is not a function " 

**Does this PR introduce a breaking change?**
no

**Other information**

### CODE REVIEW

**Comment by EugeneHlushko** [comment]
Yep no problem, where do you put the tests for options @sokra ? i couldn't see from quick look.

**Comment by sokra** [comment]
Could you check if it's possible to add these checks into the schema?

https://github.com/webpack/webpack/blob/7bbf31e8f488524e4bdd051b2a7d569560bb3ec1/schemas/webpackOptionsSchema.json#L601-L604

tests would be in test\Validation.test.js then.

File: test/Validation.test.js [hunk 1]
```diff
@@ -255,6 +255,74 @@ describe("Validation", () => {
...
 		}
+	}, {
+		name: "Invalid plugin provided: bool",
+		config: {
+			entry: "foo.js",
+			plugins: [
+				false
+			]
+		},
+		test(err) {
```
Comment (sokra) [inline test/Validation.test.js]: Could you use `message` instead of `test` to test the complete message?

Like all other tests here except one.

Comment (EugeneHlushko) [reply test/Validation.test.js]: yeah sure, nice catch :)

**Comment by EugeneHlushko** [comment]
Any action point on this regarding `webpack 4 milestone`: Do i need to rebase changes on `next` branch and change target from `master`?

File: schemas/webpackOptionsSchema.json [hunk 2]
```diff
@@ -1,6 +1,34 @@
 {
   "additionalProperties": false,
   "definitions": {
+    "common.pluginFunction": {
+      "description": "Compiler or Resolver Plugin instance of Function",
```
Comment (sokra) [inline schemas/webpackOptionsSchema.json]: Function acting as plugin

File: schemas/webpackOptionsSchema.json [hunk 3]
```diff
@@ -1,6 +1,34 @@
...
+      "instanceof": "Function",
+      "properties": {
```
Comment (sokra) [inline schemas/webpackOptionsSchema.json]: I guess this is not needed as `apply` is inherited from the `Function.prototype`.

File: schemas/webpackOptionsSchema.json [hunk 4]
```diff
@@ -1,6 +1,34 @@
...
+          "instanceof": "Function"
+        }
+      },
+      "additionalProperties": true,
+      "required": [
+        "apply"
+      ]
+    },
+    "common.pluginObject": {
+      "description": "Compiler or Resolver Plugin instance of Function",
```
Comment (sokra) [inline schemas/webpackOptionsSchema.json]: Plugin instance

File: test/Validation.test.js [hunk 5]
```diff
@@ -255,6 +255,98 @@ describe("Validation", () => {
...
+			"      -> Compiler or Resolver Plugin instance of Function",
+			"    * configuration.plugins[0] should be an instance of function",
+			"      -> Compiler or Resolver Plugin instance of Function"
+		]
+	}, {
+		name: "Invalid plugin provided: object without apply function",
+		config: {
+			entry: "foo.js",
+			plugins: [
+				new function() {}
```
Comment (sokra) [inline test/Validation.test.js]: This looks weird. Why not just passing `{}`?

Comment (EugeneHlushko) [reply test/Validation.test.js]: oh yeah this was used back on functional check as experiment, lemme update that fellow

File: schemas/webpackOptionsSchema.json (hunk 3 above)
Comment (EugeneHlushko) [reply schemas/webpackOptionsSchema.json]: this is needed or providing `{}` or `new function() {}` starts passing validation

File: schemas/webpackOptionsSchema.json [hunk 6]
```diff
@@ -1178,7 +1217,18 @@
     },
     "plugins": {
       "description": "Add additional plugins to the compiler.",
-      "type": "array"
+      "type": "array",
+      "items": {
+        "description": "Plugin of type object or instanceof Function",
```
Comment (EugeneHlushko) [inline schemas/webpackOptionsSchema.json]: good here @sokra  ?

**Comment by sokra** [comment]
Thanks

**Comment by EugeneHlushko** [comment]
👍

**Pull request merged**